src/mempool_limit.h

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2019 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef MEMPOOLLIMIT_H
+#define MEMPOOLLIMIT_H
+
+#include <map>
+#include <set>
+#include <vector>
+
+#include "boost/optional.hpp"
+#include "primitives/transaction.h"
+#include "uint256.h"
+
+const size_t DEFAULT_MEMPOOL_TOTAL_COST_LIMIT = 80000000;
+const int64_t DEFAULT_MEMPOOL_EVICTION_MEMORY_MINUTES = 60;
+
+const size_t RECENTLY_EVICTED_SIZE = 40000;
+const uint64_t MIN_TX_WEIGHT = 4000;
+const uint64_t LOW_FEE_PENALTY = 16000; 
+
+class RecentlyEvictedList
+{
+    const size_t capacity;
+
+    const int64_t timeToKeep;
+    // Pairs of txid and time (seconds since epoch)
+    std::deque<std::pair<uint256, int64_t>> txIdsAndTimes;
+    std::set<uint256> txIdSet;
+
+    void pruneList();
+
+public:
+    RecentlyEvictedList(size_t capacity_, int64_t timeToKeep_) : capacity(capacity_), timeToKeep(timeToKeep_) 
+    {
+        assert(capacity <= RECENTLY_EVICTED_SIZE);
+    }
+    RecentlyEvictedList(int64_t timeToKeep_) : RecentlyEvictedList(RECENTLY_EVICTED_SIZE, timeToKeep_) {}
+
+    void add(const uint256& txId);
+    bool contains(const uint256& txId);
+};
+
+
+struct TxWeight {
+    int64_t cost;
+    int64_t evictionWeight;
+
+    TxWeight(int64_t cost_, int64_t evictionWeight_)
+        : cost(cost_), evictionWeight(evictionWeight_) {}
+
+    TxWeight add(const TxWeight& other) const;
+    TxWeight negate() const;
+};
+
+
+struct WeightedTxInfo {
+    uint256 txId;
+    TxWeight txWeight;
+
+    WeightedTxInfo(uint256 txId_, TxWeight txWeight_) : txId(txId_), txWeight(txWeight_) {}
+
+    static WeightedTxInfo from(const CTransaction& tx, const CAmount& fee);
+};
+
+
+class WeightedTxTree
+{
+    const int64_t capacity;
+    size_t size = 0;
+    
+    std::vector<WeightedTxInfo> txIdAndWeights;
+    std::vector<TxWeight> childWeights;
+    std::map<uint256, size_t> txIdToIndexMap;
+
+    TxWeight getWeightAt(size_t index) const;
+    void backPropagate(size_t fromIndex, const TxWeight& weightDelta);
+    size_t findByEvictionWeight(size_t fromIndex, int64_t weightToFind) const;
+
+public:
+    WeightedTxTree(int64_t capacity_) : capacity(capacity_) {
+        assert(capacity >= 0);
+    }
+
+    TxWeight getTotalWeight() const;
+
+    void add(const WeightedTxInfo& weightedTxInfo);
+    void remove(const uint256& txId);
+
+    boost::optional<uint256> maybeDropRandom();
+};
```

Comment:
This is a cool object, with cool algorithms, but are we sure enough that the performance will be unacceptable if we just iterate the mempool (`CTxMemPool::mapTx`) each time we need to evict a tx? At most there will be 20k transactions, and on average we'll visit half of those. Is the overhead of iterating 10k transactions small compared with the overhead of actually adding and removing a transaction? If so, then the iteration overhead wouldn't be significant.

If the iteration overhead is significant, here's another idea I just thought of: We could evict (say) 10 transactions at a time. That introduces some error, but 10 is a tiny fraction of 20k. (It's true that we could be at the weight limit with fewer transactions, but 10 would still be a small fraction.) This doesn't conform to the https://github.com/zcash/zips/pull/275, but would the deviation be acceptable?

Here's a way to evict 10 at a time with a single iteration pass. Generate a vector of 10 random numbers between 1 and the total current weight (or one less than these values). Sort the vector low-to-high, then calculate differences. So `[2, 5, 14]` (after sorting) would become `[2, 3, 9]`. Start iterating the mempool, decrementing the first value in the vector by each transaction's weight, and if it goes negative, then evict this transaction (or add it to a temporary eviction list if we can't evict it without breaking the iterator). Discard the first value in the vector (probably by just incrementing the index). Continue iterating and evicting.

A minor variation would be to not evict these 10 immediately, but only save a list of their tx IDs. Then, when adding a new tx that would exceed the weight limit, evict the next tx on the list, if it still exists. (This would not conform to the zip in that the new tx wouldn't be eligible for eviction, but perhaps that would be acceptable.) When the list is empty, iterate to create a list of 10 more.

The nice thing about this approach is we wouldn't need an entirely new data structure; the existing `mapTx` would be sufficient.

Reply:
I think that your right in that this could be done by just iterating over `mapTx`, recalculating transaction sizes as we go until we find the transaction that we want to evict. We would also have to keep track of the the total size of the mempool as transactions are added and removed, but that shouldn't be too difficult. I like the simplicity of that solution and I wish I had thought of this when I first started working on this. In light of your suggestion, it is likely that current implementation is prematurely optimized. I haven't done any performance analysis but my intuition tells me that iterating over `mapTx` would be sufficient, but also that my current solution is more performant. That being said, the 2.1.0 release is blocked on solving this issue and I am hesitant to redo this. 🤷‍♂